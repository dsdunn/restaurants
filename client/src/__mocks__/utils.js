export const getRestaurantsAndGenres = jest.fn(() => ({
  formattedRestaurants: [],
  genres: []
}));

export const filter = jest.fn(() => {
  return [];
})