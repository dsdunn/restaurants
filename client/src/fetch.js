let key = 'q3MNxtfep8Gt';

const fetchData = async (endpoint, method = 'GET') => {
  let response = await fetch(`https://code-challenge.spectrumtoolbox.com/api/${endpoint}`, {
    headers: {
      'Authorization': `Api-Key ${key}`
    }
  });
  return await response.json();
}

export const getRestaurants = () => {
  return fetchData('restaurants');
};