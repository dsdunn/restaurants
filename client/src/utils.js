export const filter = (restaurants, criteria = {}) => {
  return restaurants.sort((jointA, jointB) => {
    return jointA.name > jointB.name ? 1 : -1;
  })
}