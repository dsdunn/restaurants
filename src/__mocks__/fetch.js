export const getRestaurants = jest.fn(() => {
  return Promise.resolve();
});