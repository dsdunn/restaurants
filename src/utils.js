export const filter = (restaurants, criteria = {}) => {
  if ( Object.values(criteria).find(criterion => criterion)) {
    restaurants = restaurants.filter(joint => {
      return matchesCriteria(joint, criteria);
    })
  }

  return restaurants;
};

export const getRestaurantsAndGenres = (restaurants = []) => {
  let genres = [];
  let genreReducer = (genreList) => {
    genreList.forEach(genre => {
      if (!genres.includes(genre)){
        genres.push(genre);
      }
    })
  }

  restaurants.forEach(joint => {
    let { genre } = joint;

    if (genre) {
      let genreList = genre.split(',');

      genreReducer(genreList);

      joint.genre = genreList.sort().join(', ');
    }
  })

  restaurants.sort((jointA, jointB) => {
    return jointA.name > jointB.name ? 1 : -1;
  });
  
  return ({
    formattedRestaurants: restaurants,
    genres: genres.sort()
  });
}

const matchesCriteria = (joint, criteria) => {
  let qualifies = true;
  let keys = Object.keys(criteria);

  for (let i = 0; i < keys.length; i++) {
    let criterion = keys[i];
    let value = criteria[criterion];

    if (!criteria[criterion]) continue;

    if (criterion === 'terms' && checkNameCityGenre(joint, criteria[criterion])) {
      continue;
    } else if (criterion !== 'terms' && joint[criterion].includes(value)) {
      continue;
    } else { 
      qualifies = false;
    }
  }
  return qualifies;
}

const checkNameCityGenre =  (joint, term) => {
  term = term.toLowerCase();

  return joint.name.toLowerCase().includes(term) || joint.city.toLowerCase().includes(term) || joint.genre.includes(term);
}




