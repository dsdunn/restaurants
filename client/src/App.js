import React, { useState, useEffect } from 'react';
import { Box, Container, Typography } from '@material-ui/core';
import { Pagination } from '@material-ui/lab';

import { FilterForm } from './components/FilterForm';
import { RestaurantsTable } from './components/RestaurantsTable';

import { getRestaurants } from './fetch';
import { filter, getGenres } from './utils';

function App() {
  const [ allRestaurants, setAllRestaurants ] = useState([]);
  const [ filteredRestaurants, setFilteredRestaurants ] = useState([]);
  const [ genres, setGenres ] = useState([]);
  const [ criteria, setCriteria ] = useState({
    terms: '',
    state: '',
    genre: ''
  });
  const [ page, setPage ] = useState(1);

  useEffect(() => {
    const init = async() => {
      let restaurants = await getRestaurants();
      let { formattedRestaurants, genres } = getGenres(restaurants);

      setAllRestaurants(formattedRestaurants);
      setGenres(genres);
    }

    init();
  }, [])

  useEffect(() => {
    let filtered = filter(allRestaurants, criteria)

    setFilteredRestaurants(filtered);
  })

  return (
    <Box>
      <Container className="App">
        <Typography variant="h1" align="center" gutterBottom >Restaurants</Typography>
        <FilterForm criteria={criteria} setCriteria={setCriteria} genres={genres}/>
        <div className="pagination">
          <Pagination count={3}/>
        </div>
        <RestaurantsTable filteredRestaurants={filteredRestaurants} />
      </Container>
    </Box>
  );
}

export default App;
