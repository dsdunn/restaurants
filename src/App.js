import React, { useState, useEffect } from 'react';
import { Box, Container, Typography } from '@material-ui/core';

import { FilterForm } from './components/FilterForm';
import { RestaurantsTable } from './components/RestaurantsTable';

import { getRestaurants } from './fetch';
import { filter, getRestaurantsAndGenres } from './utils';

function App() {
  const [ allRestaurants, setAllRestaurants ] = useState([]);
  const [ filteredRestaurants, setFilteredRestaurants ] = useState([]);
  const [ genres, setGenres ] = useState([]);
  const [ criteria, setCriteria ] = useState({
    terms: '',
    state: '',
    genre: ''
  });

  useEffect(() => {
    const init = async() => {
      let restaurants = await getRestaurants();
      let { formattedRestaurants, genres } = getRestaurantsAndGenres(restaurants);

      setAllRestaurants(formattedRestaurants);
      setFilteredRestaurants(formattedRestaurants);
      setGenres(genres);
    }

    init();
  }, [])

  useEffect(() => {
    let filtered = filter(allRestaurants, criteria)

    setFilteredRestaurants(filtered);
  }, [criteria])

  return (
    <Box>
      <Container className="App">
        <Typography variant="h1" align="center" gutterBottom >Restaurants</Typography>
        <FilterForm criteria={criteria} setCriteria={setCriteria} genres={genres}/>
        <RestaurantsTable filteredRestaurants={filteredRestaurants} />
      </Container>
    </Box>
  );
}

export default App;
