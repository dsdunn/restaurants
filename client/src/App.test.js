import React from 'react';
import { render, cleanup, act, screen } from '@testing-library/react';
import { getRestaurants } from './fetch';
import { filter, getRestaurantsAndGenres } from './utils';

import App from './App';

jest.mock('./fetch');
jest.mock('./utils');

// getRestaurantsAndGenres.mockResolvedValueOnce({restaurants: [], genres: []})

// const mockGetRestaurants = jest.fn(() => Promise.resolve([{ data: 'some data'}]));

describe('<App>', () => {
  let promise;

  beforeEach(() => {
    promise = Promise.resolve()
  })

  afterEach(() => {
    cleanup;
  });

  it('should match the snapshot', async () => {
    const { asFragment } = render( <App />);

    expect(asFragment(<App />)).toMatchSnapshot();
    await act(() => promise);
  });

  it('should call getRestuarants', async () => {
    render( <App />);
    expect(getRestaurants).toBeCalled();
    await act(() => promise);
  })

  it('should call getRestuarantsAndGenres', async () => {
    render( <App />);
    expect(getRestaurantsAndGenres).toBeCalled();

    await act(() => promise);
  })

  it('should render the FilterForm', async () => {
    render( <App />);
    const element = screen.getByTestId('filter-form');

    expect(element).toBeTruthy();
    await act(() => promise);
  })

  it('should render the Restaurants Table', async () => {
    render( <App />);
    const element = screen.getByTestId('restaurants-table');

    expect(element).toBeTruthy();
    await act(() => promise);
  })
});