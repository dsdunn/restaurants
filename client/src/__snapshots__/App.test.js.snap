// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <div
      class="MuiContainer-root App MuiContainer-maxWidthLg"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1 MuiTypography-gutterBottom MuiTypography-alignCenter"
      >
        Restaurants
      </h1>
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <form
          class="makeStyles-form-41"
          data-testid="filter-form"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-input-43"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
            >
              search
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                name="terms"
                placeholder="name/city/genre"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-130 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-132"
                >
                  <span>
                    search
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div>
            <div
              class="MuiFormControl-root makeStyles-select-42"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
                data-shrink="true"
                id="state-input-label"
              >
                State
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <div
                  aria-haspopup="listbox"
                  aria-labelledby="state-input-label state-input"
                  class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                  id="state-input"
                  role="button"
                  tabindex="0"
                >
                  All
                </div>
                <input
                  data-testid="state-input"
                  name="state"
                  type="hidden"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiFormControl-root makeStyles-select-42"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
                data-shrink="true"
                id="genre-input-label"
              >
                Genre
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <div
                  aria-haspopup="listbox"
                  aria-labelledby="genre-input-label genre-input"
                  class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                  id="genre-input"
                  role="button"
                  tabindex="0"
                >
                  All
                </div>
                <input
                  name="genre"
                  type="hidden"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-input-43"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Search
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
      <div
        data-testid="restaurants-table"
      >
        <div
          class="pagination"
        />
        <div
          class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <table
            class="MuiTable-root"
          >
            <thead
              class="MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  scope="col"
                >
                  Name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  scope="col"
                >
                  Location
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  scope="col"
                >
                  Phone
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  scope="col"
                >
                  Genres
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  colspan="4"
                >
                  Sorry, no restaurants match your search criteria
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
